services/device-registry-jdbc/src/main/java/org/eclipse/hono/deviceregistry/jdbc/impl/CredentialsServiceImpl.java

```diff
@@ -0,0 +1,150 @@
+
+import io.opentracing.Span;
+import io.vertx.core.Future;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * Implementation of the <em>Credentials Service</em>.
+ */
+public class CredentialsServiceImpl extends AbstractCredentialsService {
+
+    private final TableAdapterStore store;
+
+    /**
+     * Create a new instance.
+     *
+     * @param store The backing service to use.
+     */
+    public CredentialsServiceImpl(final TableAdapterStore store) {
+        this.store = store;
+    }
+
+    @Override
+    protected Future<CredentialsResult<JsonObject>> processGet(final TenantKey tenant, final CredentialKey key, final JsonObject clientContext, final Span span) {
+        return this.store.findCredentials(key, span.context())
+                .map(r -> {
+
+                    if (r.isEmpty()) {
+                        return CredentialsResult.from(HttpURLConnection.HTTP_NOT_FOUND);
+                    }
+
+                    final var result = r.get();
+
+                    final var secrets = result.getCredentials()
+                            .stream()
+                            .map(JsonObject::mapFrom)
+                            .filter(filter(key.getType(), key.getAuthId()))
+                            .flatMap(c -> c.getJsonArray(CredentialsConstants.FIELD_SECRETS)
+                                    .stream()
+                                    .filter(JsonObject.class::isInstance)
+                                    .map(JsonObject.class::cast))
+                            .filter(CredentialsServiceImpl::filterSecrets)
+                            .collect(Collectors.toList());
+
+                    if (secrets.isEmpty()) {
+                        // nothing was left after filtering ... not found
+                        return CredentialsResult.from(HttpURLConnection.HTTP_NOT_FOUND);
+                    }
+
+                    final var payload = new JsonObject()
+                            .put(Constants.JSON_FIELD_DEVICE_ID, result.getDeviceId())
+                            .put(CredentialsConstants.FIELD_TYPE, key.getType())
+                            .put(CredentialsConstants.FIELD_AUTH_ID, key.getAuthId())
+                            .put(CredentialsConstants.FIELD_SECRETS, new JsonArray(secrets));
+
+                    return CredentialsResult.from(HttpURLConnection.HTTP_OK, payload, CacheDirective.noCacheDirective());
+
+                });
+    }
+
+    private static boolean filterSecrets(final JsonObject secret) {
+        if (secret == null) {
+            return false;
+        }
+
+        if (!secret.getBoolean(CredentialsConstants.FIELD_ENABLED, true)) {
+            return false;
+        }
+
+        if (!validTime(secret, CredentialsConstants.FIELD_SECRETS_NOT_BEFORE, Instant::isAfter)) {
+            return false;
+        }
+        if (!validTime(secret, CredentialsConstants.FIELD_SECRETS_NOT_AFTER, Instant::isBefore)) {
```

Comment:
@sophokles73 @ctron 

Sorry for raising this comment now. I noticed this just now with the merge of the PR #2241 for adjusting the tests.

The Credentials API docs says:

> Adapters MUST only consider secrets for authentication which
> 
>     have their not-before property set to either null or the current or a past point in time and
>     have their not-after property set to either null or the current or a future point in time.

AFAIK the credentials API includes all the secrets irrespective of the _not-before_ & _not-after_ property. Based on the above quotes, the adapters are responsible for filtering them out. Various _AuthProvider_ implementations already do this by using `CredentialsObject.getCandidateSecrets(...)` where the secrets are already being filtered out based on the validity period. IMHO we need not filter secrets based on the validity period in the credentials API implementation. I don't think we need to adapt the tests as in #2241